src/txrequest.h

```diff
@@ -0,0 +1,213 @@
+ * - The peer that announced them was not overloaded.
+ * - No announcement for the same txhash from another peer within the same preferred/nonpreferred class already has
+ *   the first marker already.
+ *
+ *   Rationale: these restrictions avoid giving the speed benefit to honest but overloaded peers, and also
+ *              reduce the extent to which attackers that race the network in announcing all transactions
+ *              can break chains of dependent transactions.
+ *
+ * Together these rules strike a balance between being fast in non-adverserial conditions and avoiding influenceable
+ * by attackers. An invblocking attacker that races the network:
+ * - Will be unsuccessful if all preferred connections are honest (and there is at least one).
+ * - If there are P preferred connections of which Ph>=1 are honest, the attacker can delay us from learning
+ *   about a transaction by k expiration periods, where k ~ 1 + NHG(N=P-1,K=P-Ph-1,r=1), which has mean
+ *   P/(Ph+1) (where NHG stands for Negative Hypergeometric distribution).
+ * - If all P preferred connections are to the attacker, and there are NP non-preferred connections of which NPh are
+ *   honest, k ~ P + 1 + NHG(N=NP-1,K=NP-NPh-1,r=1), with mean P + NP/(NPh+1).
+ *
+ * Complexity:
+ * - Memory usage is proportional to the total number of tracked announcements (Size()) plus the number of
+ *   peers with a nonzero number of tracked announcements.
+ * - CPU usage is generally logarithmic in the total number of tracked announcements, plus the number of
+ *   announcements affected by an operation (amortized O(1) per announcement).
+ */
+class TxRequestTracker {
+public:
+    /** A functor with embedded salt that computes priority of a txhash/peer combination. Lower priorities are
+     *  selected first. */
+    class PriorityComputer {
+        const uint64_t m_k0, m_k1;
+    public:
+        explicit PriorityComputer(bool deterministic);
+        uint64_t operator()(const uint256& txhash, uint64_t peer, bool preferred, bool first) const;
+    };
+
+private:
+    // Avoid littering this header file with implementation details.
+    class Impl;
+    std::unique_ptr<Impl> m_impl;
+
+public:
+    //! Construct a TxRequestTracker.
+    TxRequestTracker(bool deterministic = false);
+    ~TxRequestTracker();
+
+    // Conceptually, the data structure consists of a collection of entries, one for each peer/txhash combination
+    // (an "announcement"):
+    //
+    // - CANDIDATE entries represent transactions that were announced by a peer, and that become available for
+    //   download after their reqtime has passed.
+    //
+    // - REQUESTED entries represent transactions that have been requested, and which we're awaiting a response for
+    //   from that peer. Their expiry value determines when the request times out.
+    //
+    // - COMPLETED entries represent transactions that have been requested from a peer, but they timed out, a
```

Comment:
`COMPLETED` sounds more like `FAILED`.

`COMPLETED` is also hit when the transaction is received properly, no?

Reply:
I'm not sure what you're suggesting here. FAILED seems like a bad name as it indeed doesn't imply failure.

Reply:
apologies I wrote my comment in two frame of minds. I mean to say the comment doesn't mention "success" as a possibility to transition to this state. Update the comment and I'm happy.

Reply:
Done. Better?

Reply:
yes, thank you